CDS/src/org/icpc/tools/cds/service/ContestRESTService.java

```diff
@@ -635,6 +641,149 @@ else if (!IContestObject.isSingleton(cType)) {
+		if (segments == null || segments.length != 2) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+
+		ConfiguredContest cc = CDSConfig.getContest(segments[0]);
+		if (cc == null) {
+			response.sendError(HttpServletResponse.SC_NOT_FOUND, "Contest not found");
+			return;
+		}
+
+		String type = segments[1];
+		ContestType cType = IContestObject.getTypeByName(type);
+		if (cType != ContestType.SUBMISSION) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid type");
+			return;
+		}
+
+		JsonObject obj = null;
+		try {
+			InputStream is = request.getInputStream();
+			JSONParser parser = new JSONParser(is);
+			obj = parser.readObject();
+		} catch (Exception e) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Could not parse object");
+			return;
+		}
+
+		String id = obj.getString("id");
+		String time = obj.getString("time");
+		String teamId = obj.getString("team_id");
+		String problemId = obj.getString("problem_id");
+		String languageId = obj.getString("language_id");
+
+		Contest contest = cc.getContestByRole(request);
+		if (teamId != null && contest.getTeamById(teamId) == null) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Unknown team");
+			return;
+		}
+
+		if (languageId == null || contest.getLanguageById(languageId) == null) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Unknown language");
+			return;
+		}
+
+		if (problemId == null || contest.getProblemById(problemId) == null) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Unknown problem");
+			return;
+		}
+
+		if (Role.isTeam(request)) {
+			if (id != null || time != null) {
+				response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Team cannot assign id or time");
+				return;
+			}
+			String teamId2 = CDSConfig.getInstance().getTeamIdFromUser(request.getRemoteUser());
+			if (teamId2 == null) {
+				response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Could not determine request user's team");
+				return;
+			}
+			if (teamId != null && !teamId.equals(teamId2)) {
+				response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Cannot submit for a different team");
+				return;
+			}
+			// make sure team_id is not null in case we're forwarding to CCS
+			if (teamId == null)
+				obj.props.put("team_id", teamId2);
+		}
+
+		if (Role.isAdmin(request) && teamId == null) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST, "No team specified");
+			return;
+		}
+
+		String sId = null;
+		ContestSource source = cc.getContestSource();
+
+		if (cc.isTesting()) {
+			// when in test mode just accept submissions and return dummy id
+			sId = "test-" + obj.getString("team_id");
+		} else if (cc.getCCS() == null) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST, "No CCS configured");
+			return;
+		} else if (!(source instanceof RESTContestSource)) {
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST, "CCS does not support submissions via the CCS");
```

Comment:
Strange line?

Reply:
Ah, second instance should be CDS. This gives an error if the CCS is configured via XML event feed and hence doesn't support submissions via the CDS. Fixed via a commit.